services/libs/integrations/src/integrations/github/processStream.ts

```diff
@@ -0,0 +1,844 @@
+import { getAppToken } from './api/rest/getAppToken'
+import { createAppAuth } from '@octokit/auth-app'
+import { AuthInterface } from '@octokit/auth-app/dist-types/types'
+
+const IS_TEST_ENV: boolean = process.env.NODE_ENV === 'test'
+
+let githubAuthenticator: AuthInterface | undefined = undefined
+
+function getAuth(ctx: IProcessStreamContext): AuthInterface | undefined {
+  const GITHUB_CONFIG = ctx.platformSettings as GithubPlatformSettings
+  const privateKey = GITHUB_CONFIG.privateKey
+    ? Buffer.from(GITHUB_CONFIG.privateKey, 'base64').toString('ascii')
+    : undefined
+
+  if (githubAuthenticator === undefined) {
+    githubAuthenticator = privateKey
+      ? createAppAuth({
+          appId: GITHUB_CONFIG.appId,
+          clientId: GITHUB_CONFIG.clientId,
+          clientSecret: GITHUB_CONFIG.clientSecret,
+          privateKey,
+        })
+      : undefined
+  }
+  return githubAuthenticator
+}
+
+const getTokenFromCache = async (ctx: IProcessStreamContext) => {
+  const key = 'github-token-cache'
+  const cache = ctx.cache
+
+  const token = await cache.get(key)
+
+  if (token) {
+    return JSON.parse(token)
+  } else {
+    return null
+  }
+}
+
+const setTokenToCache = async (ctx: IProcessStreamContext, token: AppTokenResponse) => {
+  const key = 'github-token-cache'
+  const cache = ctx.cache
+
+  // cache for 5 minutes
+  await cache.set(key, JSON.stringify(token), 5 * 60)
+}
+
+async function getGithubToken(ctx: IProcessStreamContext): Promise<string> {
+  const auth = getAuth(ctx)
+  if (auth) {
+    let appToken: AppTokenResponse
+    const tokenFromCache = await getTokenFromCache(ctx)
+    if (tokenFromCache) {
+      appToken = tokenFromCache
+    } else {
+      // no token or it expired (more 5 minutes)
+      const authResponse = await auth({ type: 'app' })
+      const jwtToken = authResponse.token
+      appToken = await getAppToken(jwtToken, ctx.integration.identifier)
+    }
+
+    setTokenToCache(ctx, appToken)
+
+    return appToken.token
+  }
+
+  throw new Error('GitHub integration is not configured!')
+}
+
+async function getMemberData(ctx: IProcessStreamContext, login: string): Promise<any> {
+  const appToken = await getGithubToken(ctx)
+  return getMember(login, appToken)
+}
+
+async function getMemberEmail(ctx: IProcessStreamContext, login: string): Promise<string> {
+  if (IS_TEST_ENV) {
+    return ''
+  }
+
+  const cache = ctx.cache
+
+  const existing = await cache.get(login)
+  if (existing) {
+    if (existing === 'null') {
+      return ''
+    }
+
+    return existing
+  }
+
+  const member = await getMemberData(ctx, login)
+  const email = (member && member.email ? member.email : '').trim()
+  if (email && email.length > 0) {
+    await cache.set(login, email, 60 * 60)
+    return email
+  }
+
+  await cache.set(login, 'null', 60 * 60)
+  return ''
+}
+
+const handleNextPageStream = async (ctx: IProcessStreamContext, response: GraphQlQueryResponse) => {
```

Comment:
can we give this a better name? maybe `publishNextPageStream`

Reply:
Updated